File: sdk/beacon_sdk/tracer.py
from __future__ import annotations

import logging
import uuid
from contextlib import contextmanager
from contextvars import Token
from typing import Any, Generator

from beacon_sdk.context import (
    TraceContext,
    get_context,
    register_span,
    reset_context,
    set_context,
    unregister_span,
)
from beacon_sdk.exporters import SpanExporter
from beacon_sdk.models import Span, SpanStatus, SpanType

logger = logging.getLogger("beacon_sdk")


class BeaconTracer:
    """Creates and manages spans within a trace context."""

    def __init__(
        self,
        exporter: SpanExporter | None = None,
        enabled: bool = True,
    ) -> None:
        self._exporter = exporter
        self._enabled = enabled

    def start_span(
        self,
        name: str,
        span_type: SpanType = SpanType.CUSTOM,
        attributes: dict[str, Any] | None = None,
    ) -> tuple[Span, Token[TraceContext | None]]:
        """Start a new span. Returns (span, context_token)."""
        current_ctx = get_context()

        if current_ctx is not None:
            trace_id = current_ctx.trace_id
            parent_span_id = current_ctx.span_id
        else:
            trace_id = str(uuid.uuid4())
            parent_span_id = None

        span = Span(
            trace_id=trace_id,
            parent_span_id=parent_span_id,
            span_type=span_type,
            name=name,
            attributes=attributes or {},
        )

        new_ctx = TraceContext(trace_id=trace_id, span_id=span.span_id)
        token = set_context(new_ctx)
        register_span(span)

        return span, token

    def end_span(
        self,
        span: Span,
        token: Token[TraceContext | None],
        status: SpanStatus = SpanStatus.OK,
        error_message: str | None = None,
    ) -> None:
        """End a span, restore context, and export."""
        try:
            span.end(status=status, error_message=error_message)
        except Exception as exc:
            logger.debug("Beacon: failed to finalize span: %s", exc)
        finally:
            reset_context(token)
            unregister_span(span.span_id)

        if self._enabled and self._exporter is not None:
            try:
                self._exporter.export([span])
            except Exception as exc:
                logger.debug(
                    "Beacon: span export failed (non-fatal): %s", exc
                )

    @contextmanager
    def span(
        self,
        name: str,
        span_type: SpanType = SpanType.CUSTOM,
        attributes: dict[str, Any] | None = None,
    ) -> Generator[Span, None, None]:
        """Context manager for spans."""
        span, token = self.start_span(
            name, span_type=span_type, attributes=attributes
        )
        try:
            yield span
            self.end_span(span, token, status=SpanStatus.OK)
        except Exception as exc:
            self.end_span(
                span, token, status=SpanStatus.ERROR, error_message=str(exc)
            )
            raise

File: sdk-js/src/tracer.ts
/**
 * BeaconTracer — core tracing engine.
 *
 * Mirrors Python SDK's beacon_sdk/tracer.py but uses AsyncLocalStorage
 * for context propagation instead of ContextVar tokens.
 */

import { randomUUID } from "node:crypto";
import {
  getContext,
  runWithContext,
  registerSpan,
  unregisterSpan,
  type TraceContext,
} from "./context.js";
import type { SpanExporter } from "./exporter.js";
import { Span, SpanStatus, type SpanType } from "./models.js";

export class BeaconTracer {
  private readonly exporter: SpanExporter | null;
  private readonly enabled: boolean;

  constructor(options: {
    exporter: SpanExporter | null;
    enabled?: boolean;
  }) {
    this.exporter = options.exporter;
    this.enabled = options.enabled ?? true;
  }

  startSpan(
    name: string,
    options?: {
      spanType?: SpanType;
      attributes?: Record<string, unknown>;
    }
  ): Span {
    const currentCtx = getContext();
    const traceId = currentCtx?.traceId ?? randomUUID();
    const parentSpanId = currentCtx?.spanId ?? null;

    const span = new Span({
      trace_id: traceId,
      parent_span_id: parentSpanId,
      span_type: options?.spanType,
      name,
      attributes: options?.attributes,
    });

    registerSpan(span);
    return span;
  }

  endSpan(
    span: Span,
    status: SpanStatus = SpanStatus.OK,
    errorMessage?: string
  ): void {
    try {
      span.end(status, errorMessage);
    } finally {
      unregisterSpan(span.span_id);
    }

    if (this.enabled && this.exporter) {
      try {
        this.exporter.export([span]);
      } catch {
        // Silent failure, matching Python SDK behavior
      }
    }
  }

  /**
   * Run a function within a span context.
   * Handles both sync and async functions with automatic context propagation.
   */
  withSpan<T>(
    name: string,
    options: {
      spanType?: SpanType;
      attributes?: Record<string, unknown>;
    },
    fn: (span: Span) => T
  ): T {
    const span = this.startSpan(name, options);
    const ctx: TraceContext = { traceId: span.trace_id, spanId: span.span_id };

    return runWithContext(ctx, () => {
      try {
        const result = fn(span);
        if (result instanceof Promise) {
          return result
            .then((val: unknown) => {
              this.endSpan(span, SpanStatus.OK);
              return val;
            })
            .catch((err: unknown) => {
              this.endSpan(
                span,
                SpanStatus.ERROR,
                err instanceof Error ? err.message : String(err)
              );
              throw err;
            }) as T;
        }
        this.endSpan(span, SpanStatus.OK);
        return result;
      } catch (err) {
        this.endSpan(
          span,
          SpanStatus.ERROR,
          err instanceof Error ? err.message : String(err)
        );
        throw err;
      }
    });
  }
}
